PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+// This file is part of Notepad++ project
+// Copyright (C)2003 Don HO <don.h@free.fr>
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either
+// version 2 of the License, or (at your option) any later version.
+//
+// Note that the GPL places important restrictions on "derived works", yet
+// it does not provide a detailed definition of that term.  To avoid      
+// misunderstandings, we consider an application to constitute a          
+// "derivative work" for the purpose of this license if it does any of the
+// following:                                                             
+// 1. Integrates source code from Notepad++.
+// 2. Integrates/includes/aggregates Notepad++ into a proprietary executable
+//    installer, such as those produced by InstallShield.
+// 3. Links to a library or executes a program that does any of the above.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+#include "Directory.h"
+#include <Shlwapi.h>
+
+Directory::Directory(bool hideEmptyDirs)
+	: _exists(false)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	_filters.emplace_back(generic_string(TEXT("*.*")));
+}
+
+Directory::Directory(const generic_string& path, const std::vector<generic_string>& filters, bool hideEmptyDirs, bool autoread)
+	: _exists(false)
+	, _path(path)
+	, _filters(filters)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	if (autoread)
+		read(path);
+}
+
+bool Directory::read(const generic_string& path, const std::vector<generic_string>& filters)
+{
+	_path = path;
+	_exists = false;
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	_files.clear();
+	_dirs.clear();
+	_invisibleDirs.clear();
+
+	if (&_filters != &filters)
+		_filters = filters;
+
+	// first, read directories, never filtered.
+	append(_path, TEXT("*.*"), true);
+
+	// then read files, by each filter
+	if (_filters.empty())
+	{
+		append(_path, TEXT("*.*"), false);
+	}
+	else
+	{
+		for (size_t i=0; i<_filters.size(); ++i)
+			append(_path, _filters[i], false);
+	}
+
+
+	readLastWriteTime(_lastWriteTime);
+
+	_wasRead = true;
+
+	return _exists;
```

Comment:
I think because append() is called three times here and the return value would have to be or'ed each time, it's better to set the member variable. Additionally, it is a quick way to compare existing and non-existing directories, and can be accessed via getter.